File: abc253/c/main.py
Q = int(input())

s = set()
cnt = {}
for _ in range(Q):
    q, *num = map(str, input().split())
    if q == '1':
        s.add(int(num[0]))
        if cnt.get(num[0]) != None:
            cnt[num[0]] += 1
        else:
            cnt[num[0]] = 1
    elif q == '2':
        if cnt.get(num[0]) != None:
            cnt[num[0]] -= min(cnt[num[0]], int(num[1]))
            if cnt[num[0]] == 0:
                s.remove(int(num[0]))
    else:
        l = list(s)
        print(l)
        print(l[-1] - l[0])
        

File: abc253/c/main.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

int main() {
    set<ll> st;
    map<ll, ll> cnt;

    ll Q;
    cin >> Q;
    for (ll i = 0; i < Q; i++) {
        ll num;
        cin >> num;
        if (num == 1) {
            ll x;
            cin >> x;
            st.insert(x);
            cnt[x] += 1;
        } else if (num == 2) {
            ll x, c;
            cin >> x >> c;
            cnt[x] -= min(c, cnt[x]);
            if (cnt[x] == 0) {
                st.erase(x);
            }
        } else {
            cout << *st.rbegin() - *st.begin() << endl;
        }
    }

    return 0;
}